File: src/Chunker.php
<?php

namespace RAG;

// ============================================================
// Chunker — Split Documents Into Digestible Pieces
// ============================================================
// An embedding model has a token limit — we can't feed it an
// entire 50-page document at once. We split the document into
// overlapping chunks that are:
//
//   - Small enough for the embedding model to process
//   - Large enough to carry meaningful context
//   - Overlapping so sentences on boundaries aren't lost
//   - Prefixed with source + section so the embedding knows
//     WHERE each piece came from, not just what it says
//
// Best practice: one idea per chunk. We split on paragraph
// boundaries and respect section headings as natural dividers.
// ============================================================

class Chunker
{
    private int $chunkSize;    // max characters per chunk
    private int $overlap;      // chars carried over to next chunk
    private int $minChunkSize; // discard fragments smaller than this

    public function __construct(int $chunkSize = 1500, int $overlap = 150, int $minChunkSize = 100)
    {
        $this->chunkSize    = $chunkSize;
        $this->overlap      = $overlap;
        $this->minChunkSize = $minChunkSize;
    }

    // ----------------------------------------------------------
    // Main entry point.
    //
    // We walk through the document paragraph by paragraph.
    // When adding the next paragraph would exceed chunkSize,
    // we save the current chunk and start a new one — but carry
    // a small overlap from the previous chunk so context isn't
    // lost at the boundary.
    //
    // $source is the document name — we prefix every chunk with
    // it so the embedding knows "this is about X" not just the
    // raw paragraph text.
    // ----------------------------------------------------------
    public function chunk(string $text, string $source = ''): array
    {
        $text       = $this->cleanText($text);
        $paragraphs = preg_split('/\n{2,}/', $text);

        $chunks         = [];
        $currentText    = '';
        $currentSection = '';

        foreach ($paragraphs as $paragraph) {
            $paragraph = trim($paragraph);
            if ($paragraph === '') continue;

            // Section headings are natural chunk boundaries.
            // When we hit one, flush whatever we've built so far
            // and start fresh under the new section.
            if ($this->looksLikeHeading($paragraph)) {
                $this->flushChunk($currentText, $source, $currentSection, $chunks);
                $currentSection = $paragraph;
                $currentText    = '';
                continue;
            }

            $textIfWeAddThisParagraph = $currentText === ''
                ? $paragraph
                : $currentText . "\n\n" . $paragraph;

            // If adding this paragraph would make the chunk too big,
            // save what we have and start a new chunk with overlap
            if (mb_strlen($textIfWeAddThisParagraph) > $this->chunkSize && $currentText !== '') {
                $this->flushChunk($currentText, $source, $currentSection, $chunks);
                $overlap     = $this->extractOverlap($currentText);
                $currentText = $overlap !== '' ? $overlap . "\n\n" . $paragraph : $paragraph;
            } else {
                $currentText = $textIfWeAddThisParagraph;
            }
        }

        // Don't forget the last chunk
        $this->flushChunk($currentText, $source, $currentSection, $chunks);

        return $chunks;
    }

    // ----------------------------------------------------------
    // Save a completed chunk to the results array.
    // We skip chunks that are too short to be meaningful
    // (leftover fragments, headers without content, etc.)
    // ----------------------------------------------------------
    private function flushChunk(string $text, string $source, string $section, array &$chunks): void
    {
        $text = trim($text);

        if (mb_strlen($text) < $this->minChunkSize) return;

        $chunks[] = $this->addContextPrefix($text, $source, $section);
    }

    // ----------------------------------------------------------
    // Prefix every chunk with source + section metadata.
    //
    // WHY: Without this, the embedding only knows the raw text.
    // With this, the embedding also captures context like:
    //   "Source: onlinecheckwriter | Section: PAIN POINTS"
    //
    // This dramatically improves retrieval accuracy because
    // queries about "pain points" will match these chunks even
    // if the chunk body doesn't repeat "pain points" explicitly.
    // ----------------------------------------------------------
    private function addContextPrefix(string $text, string $source, string $section): string
    {
        $prefix = '';

        if ($source !== '')  $prefix .= "Source: $source\n";
        if ($section !== '') $prefix .= "Section: $section\n";

        return $prefix !== '' ? trim($prefix . "\n" . $text) : $text;
    }

    // ----------------------------------------------------------
    // Pull the last sentence(s) from a chunk to use as overlap.
    //
    // WHY: If chunk 1 ends mid-thought and chunk 2 starts fresh,
    // we lose the connection. By repeating the tail of chunk 1
    // at the start of chunk 2, both chunks share that bridge.
    // ----------------------------------------------------------
    private function extractOverlap(string $text): string
    {
        if ($this->overlap <= 0) return '';

        // Take the last N*3 chars and find the first sentence boundary
        $tail = mb_substr($text, -($this->overlap * 3));

        foreach (['. ', ".\n", '? ', '! '] as $boundary) {
            $position = mb_strpos($tail, $boundary);
            if ($position !== false) {
                return trim(mb_substr($tail, $position + mb_strlen($boundary)));
            }
        }

        return trim($tail);
    }

    // ----------------------------------------------------------
    // Detect section headings so we can use them as chunk dividers.
    //
    // We look for ALL CAPS lines, markdown headings (#), and
    // numbered sections. We skip long lines — headings are short.
    // ----------------------------------------------------------
    private function looksLikeHeading(string $text): bool
    {
        if (mb_strlen($text) > 120) return false;

        $isAllCaps       = $text === mb_strtoupper($text) && mb_strlen($text) > 5;
        $isMarkdown      = (bool) preg_match('/^#{1,4}\s+/', $text);
        $isNumberedOrKey = (bool) preg_match('/^(\d+\.|PAIN POINT \d+)/i', $text);

        return $isAllCaps || $isMarkdown || $isNumberedOrKey;
    }

    // ----------------------------------------------------------
    // Clean the raw document text before chunking:
    //   - Remove page numbers ("Page 1 of 10", "- 5 -")
    //   - Remove repeated separator lines (===, ---, ***)
    //   - Collapse multiple spaces/tabs into one
    //   - Reduce 3+ blank lines down to 2
    // ----------------------------------------------------------
    private function cleanText(string $text): string
    {
        $text = preg_replace('/\bpage\s+\d+(\s+of\s+\d+)?\b/i', '', $text);
        $text = preg_replace('/^\s*-\s*\d+\s*-\s*$/m', '', $text);
        $text = preg_replace('/^[=\-\*]{3,}\s*$/m', '', $text);
        $text = preg_replace('/[ \t]+/', ' ', $text);
        $text = preg_replace('/\n{3,}/', "\n\n", $text);
        return trim($text);
    }
}

File: c/chunker.c
/*
 * chunker.c — Document chunking implementation.
 *
 * The chunking algorithm works in two phases:
 *
 *   Phase 1 — Paragraph splitting.
 *   We scan the text for double-newlines (\n\n) and split there.
 *   This gives us "paragraphs" — natural semantic units in markdown
 *   and most plain-text documents.
 *
 *   Phase 2 — Size enforcement.
 *   If a paragraph exceeds chunk_size, we subdivide it at the nearest
 *   word boundary (last space before the limit).  This avoids cutting
 *   mid-word while still respecting the size target.
 *
 *   Heading detection adds a split opportunity at heading lines even
 *   inside a long paragraph.  A heading line is:
 *     - Starts with one or more '#' characters (Markdown), OR
 *     - Is entirely uppercase and short (<= 80 chars, no punctuation)
 *       (common in README / text files).
 *
 * We accumulate paragraphs into a growing "current chunk" until adding
 * the next paragraph would push us over chunk_size.  At that point we
 * flush the current chunk and start a new one.  This paragraph-coalescing
 * strategy keeps related short paragraphs together rather than creating
 * many tiny chunks.
 *
 * Memory note: every Chunk.text is a fresh malloc.  We build strings
 * with a dynamically growing buffer (chunk_buf) and strdup them when
 * flushing.  chunk_buf is stack-allocated at a generous fixed size —
 * 4 * chunk_size is always enough because we flush before the buffer
 * overflows.
 */

#include "chunker.h"
#include "config.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

/* Maximum paragraph size we'll accumulate into the per-chunk scratch buffer */
#define SCRATCH_SIZE (DEFAULT_CHUNK_SIZE * 4)

/* ---------------------------------------------------------------
 * Helper: detect whether a line looks like a heading.
 *
 * We look at just the first line of a paragraph (up to the first '\n').
 * A Markdown heading starts with '#'.
 * An all-caps heading is a short line with no lowercase letters.
 * --------------------------------------------------------------- */

static int is_heading(const char *para) {
    /* Markdown heading: starts with # */
    if (para[0] == '#') return 1;

    /* All-caps heading: find the first line, check it's short and uppercase */
    const char *nl = strchr(para, '\n');
    size_t line_len = nl ? (size_t)(nl - para) : strlen(para);

    if (line_len == 0 || line_len > 80) return 0;

    int has_lower = 0;
    int has_alpha = 0;
    for (size_t i = 0; i < line_len; i++) {
        if (islower((unsigned char)para[i])) { has_lower = 1; break; }
        if (isalpha((unsigned char)para[i]))   has_alpha = 1;
    }
    return (has_alpha && !has_lower);
}

/* ---------------------------------------------------------------
 * Helper: flush the current scratch buffer as a new Chunk.
 *
 * Trims leading/trailing whitespace, skips empty buffers.
 * Returns 1 if a chunk was added, 0 if skipped.
 * --------------------------------------------------------------- */

static int flush_chunk(char       *scratch,
                       int        *scratch_len,
                       const char *source,
                       Chunk      *chunks,
                       int        *chunk_count,
                       int         max_chunks)
{
    if (*scratch_len == 0) return 0;
    scratch[*scratch_len] = '\0';

    /* Trim trailing whitespace */
    int end = *scratch_len - 1;
    while (end >= 0 && (scratch[end] == '\n' || scratch[end] == ' ' || scratch[end] == '\t'))
        end--;

    if (end < 0) {
        *scratch_len = 0;
        return 0;
    }
    scratch[end + 1] = '\0';

    /* Trim leading whitespace */
    char *start = scratch;
    while (*start == '\n' || *start == ' ' || *start == '\t') start++;

    if (*start == '\0' || *chunk_count >= max_chunks) {
        *scratch_len = 0;
        return 0;
    }

    chunks[*chunk_count].text   = strdup(start);
    chunks[*chunk_count].source = strdup(source);
    chunks[*chunk_count].index  = *chunk_count;

    (*chunk_count)++;
    *scratch_len = 0;
    return 1;
}

/* ---------------------------------------------------------------
 * Helper: split a long paragraph at word boundaries.
 *
 * If para is longer than chunk_size, we emit it as multiple chunks,
 * each at most chunk_size characters, breaking at the last space
 * before the limit.
 *
 * This is the fallback for very long paragraphs (e.g. a wall of text
 * with no blank lines).
 * --------------------------------------------------------------- */

static int split_long_paragraph(const char *para,
                                const char *source,
                                Chunk      *chunks,
                                int        *chunk_count,
                                int         max_chunks,
                                int         chunk_size)
{
    size_t len    = strlen(para);
    size_t offset = 0;

    while (offset < len && *chunk_count < max_chunks) {
        size_t remaining = len - offset;
        size_t take      = (remaining > (size_t)chunk_size) ? (size_t)chunk_size : remaining;

        /* Find last space within the take window */
        if (offset + take < len) {
            size_t break_at = take;
            while (break_at > 0 && para[offset + break_at] != ' ') break_at--;
            if (break_at > 0) take = break_at;
        }

        /* Build chunk string */
        char *text = (char *)malloc(take + 1);
        if (!text) return -1;
        memcpy(text, para + offset, take);
        text[take] = '\0';

        chunks[*chunk_count].text   = text;
        chunks[*chunk_count].source = strdup(source);
        chunks[*chunk_count].index  = *chunk_count;
        (*chunk_count)++;

        /* Advance past whitespace */
        offset += take;
        while (offset < len && para[offset] == ' ') offset++;
    }
    return 0;
}

/* ---------------------------------------------------------------
 * chunk_text — public API.
 *
 * We split the input on \n\n to get paragraphs, then coalesce
 * short paragraphs into chunks.  Long paragraphs go through
 * split_long_paragraph.
 * --------------------------------------------------------------- */

int chunk_text(const char *text,
               const char *source,
               Chunk      *chunks,
               int         max_chunks,
               int         chunk_size)
{
    if (!text || !source || !chunks || max_chunks <= 0) return -1;

    int  chunk_count = 0;

    /*
     * Scratch buffer for the current accumulating chunk.
     * We allocate on the heap (chunk_size * 4 can be 6 KB).
     */
    int   scratch_cap = chunk_size * 4 + 4;
    char *scratch     = (char *)malloc(scratch_cap);
    if (!scratch) return -1;
    int   scratch_len = 0;

    /* Work on a copy so we can strtok it safely. */
    char *copy = strdup(text);
    if (!copy) { free(scratch); return -1; }

    /*
     * Split on \n\n.  strtok_r doesn't work well here because we want
     * to detect double-newlines specifically (strtok collapses runs of
     * delimiters).  Instead we scan manually.
     */
    char *p   = copy;
    char *end = copy + strlen(copy);

    while (p < end && chunk_count < max_chunks) {
        /* Find next \n\n */
        char *split = strstr(p, "\n\n");
        size_t para_len;
        char  *para_end;

        if (split) {
            para_len = (size_t)(split - p);
            para_end = split;
        } else {
            para_len = (size_t)(end - p);
            para_end = end;
        }

        /* Grab this paragraph */
        char para_save = *para_end;
        *para_end = '\0';
        const char *para = p;

        /* Skip empty paragraphs */
        if (para_len == 0 || strspn(para, " \t\n\r") == para_len) {
            *para_end = para_save;
            p = split ? split + 2 : end;
            continue;
        }

        /*
         * If the paragraph is itself a heading, flush whatever is
         * accumulating so the heading starts a fresh chunk.
         */
        if (is_heading(para) && scratch_len > 0) {
            flush_chunk(scratch, &scratch_len, source,
                        chunks, &chunk_count, max_chunks);
        }

        /*
         * If this paragraph alone exceeds chunk_size, flush current
         * accumulation and split the paragraph directly.
         */
        if ((int)para_len > chunk_size) {
            flush_chunk(scratch, &scratch_len, source,
                        chunks, &chunk_count, max_chunks);
            *para_end = para_save;
            /* Temporarily null-terminate the paragraph */
            char save2 = para_end[0];
            para_end[0] = '\0';
            split_long_paragraph(para, source, chunks, &chunk_count,
                                 max_chunks, chunk_size);
            para_end[0] = save2;
            p = split ? split + 2 : end;
            continue;
        }

        /*
         * Would adding this paragraph overflow the current chunk?
         * If so, flush first.
         */
        if (scratch_len + (int)para_len + 2 > chunk_size && scratch_len > 0) {
            flush_chunk(scratch, &scratch_len, source,
                        chunks, &chunk_count, max_chunks);
        }

        /* Append paragraph to scratch */
        if (scratch_len > 0) {
            scratch[scratch_len++] = '\n';
            scratch[scratch_len++] = '\n';
        }
        memcpy(scratch + scratch_len, para, para_len);
        scratch_len += (int)para_len;

        *para_end = para_save;
        p = split ? split + 2 : end;
    }

    /* Flush any remaining content */
    if (scratch_len > 0 && chunk_count < max_chunks) {
        flush_chunk(scratch, &scratch_len, source,
                    chunks, &chunk_count, max_chunks);
    }

    free(scratch);
    free(copy);
    return chunk_count;
}

/*
 * chunker_free_chunks — release Chunk.text and Chunk.source.
 *
 * We set each pointer to NULL after freeing so double-free is safe
 * (calling this function twice on the same array is a no-op).
 */
void chunker_free_chunks(Chunk *chunks, int count) {
    for (int i = 0; i < count; i++) {
        free(chunks[i].text);   chunks[i].text   = NULL;
        free(chunks[i].source); chunks[i].source = NULL;
    }
}
